engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobFailProcessor.java

```diff
@@ -54,92 +59,105 @@ public JobFailProcessor(
-      if (jobErrorMessage.capacity() > 0) {
-        incidentErrorMessage = jobErrorMessage;
-      }
-
-      incidentEvent.reset();
-      incidentEvent
-          .setErrorType(ErrorType.JOB_NO_RETRIES)
-          .setErrorMessage(incidentErrorMessage)
-          .setBpmnProcessId(value.getBpmnProcessIdBuffer())
-          .setProcessDefinitionKey(value.getProcessDefinitionKey())
-          .setProcessInstanceKey(value.getProcessInstanceKey())
-          .setElementId(value.getElementIdBuffer())
-          .setElementInstanceKey(value.getElementInstanceKey())
-          .setJobKey(key)
-          .setVariableScopeKey(value.getElementInstanceKey());
-
-      stateWriter.appendFollowUpEvent(
-          keyGenerator.nextKey(), IncidentIntent.CREATED, incidentEvent);
-    }
+    preconditionChecker
+        .check(state, jobKey)
+        .ifRightOrLeft(
+            ok -> failJob(record),
+            violation -> {
+              rejectionWriter.appendRejection(record, violation.getLeft(), violation.getRight());
+              responseWriter.writeRejectionOnCommand(
+                  record, violation.getLeft(), violation.getRight());
+            });
   }
 
-  private void acceptCommand(
-      final TypedRecord<JobRecord> command, final CommandControl<JobRecord> commandControl) {
-    final long jobKey = command.getKey();
+  private void failJob(final TypedRecord<JobRecord> record) {
+    final long jobKey = record.getKey();
+    final JobRecord failJobCommandRecord = record.getValue();
+    final var retries = failJobCommandRecord.getRetries();
+    final var retryBackOff = failJobCommandRecord.getRetryBackoff();
+
     final JobRecord failedJob = jobState.getJob(jobKey);
-    final var retries = command.getValue().getRetries();
-    final var retryBackOff = command.getValue().getRetryBackoff();
     failedJob.setRetries(retries);
     failedJob.setErrorMessage(
-        limitString(command.getValue().getErrorMessage(), MAX_ERROR_MESSAGE_SIZE));
+        limitString(failJobCommandRecord.getErrorMessage(), MAX_ERROR_MESSAGE_SIZE));
     failedJob.setRetryBackoff(retryBackOff);
-    failedJob.setVariables(command.getValue().getVariablesBuffer());
+    failedJob.setVariables(failJobCommandRecord.getVariablesBuffer());
 
     if (retries > 0 && retryBackOff > 0) {
-      final long receivedTime = command.getTimestamp();
+      final long receivedTime = record.getTimestamp();
       failedJob.setRecurringTime(receivedTime + retryBackOff);
       sideEffectWriter.appendSideEffect(
           () -> {
             jobBackoffChecker.scheduleBackOff(retryBackOff + receivedTime);
             return true;
           });
     }
-    commandControl.accept(JobIntent.FAILED, failedJob);
+    stateWriter.appendFollowUpEvent(jobKey, JobIntent.FAILED, failedJob);
+    responseWriter.writeEventOnCommand(jobKey, JobIntent.FAILED, failedJob, record);
     jobMetrics.jobFailed(failedJob.getType());
 
+    setFailedVariables(failedJob);
```

Comment:
Yes, we do. It's because we want to have these variables the next time we process the job (in case some intermediate state or error code is stored in them).

Why before the job push:
Since we may push the job immediately, it would be good that we set the new variables before they are obtained again by the job push logic.

Why before raising the incident:
The user should be able to view all the variables before the incident was raised, including the ones provided by the job failure, so it would be good to add them to the state before the incident is written to the log stream and exported to Operate, where users will be able to inspect it.